### ISSUE TITLE
Реализация подхода Energy Score для матчинга вакансий и резюме

### ISSUE THREAD

**Issue by slavb18** [iconicompany/iskillmatching#1]
Для реализации подхода **Energy Score** (энергетической оценки) на основе статьи [CALM](https://arxiv.org/abs/2510.27688), нам нужно отойти от стандартного сравнения «вектор к вектору» и начать рассматривать навыки как точки в непрерывном распределении.

В CALM энергетическая оценка используется как **likelihood-free** метрика, которая измеряет расстояние между предсказанным распределением и целевым вектором.

Вот пример реализации на Python с использованием PyTorch.

### 1. Реализация Energy Score (Метрика сходства)

Энергетическое расстояние (Energy Distance) между предсказанием  и целью  вычисляется по формуле:

$$E(P, y) = \mathbb{E}_{x \sim P} \|x - y\|_2 - \frac{1}{2} \mathbb{E}_{x, x' \sim P} \|x - x'\|_2$$

Где первая часть — это среднее расстояние от образцов до цели, а вторая — внутреннее разнообразие образцов (чтобы модель не «схлопывалась» в одну точку).

```python
import torch
import torch.nn as nn
import torch.nn.functional as F

def energy_score(target, samples):
    """
    target: [batch_size, dim] - вектор навыка из вакансии
    samples: [batch_size, num_samples, dim] - сгенерированные/извлеченные векторы из резюме
    """
    # 1. Дистанция между образцами и целью: E||x - y||
    # [batch_size, num_samples]
    dist_to_target = torch.norm(samples - target.unsqueeze(1), p=2, dim=-1).mean(dim=1)

    # 2. Внутренняя дистанция между образцами: E||x - x'|| (Diversity term)
    # Используем матричное вычисление для всех пар образцов
    s1 = samples.unsqueeze(2) # [B, N, 1, D]
    s2 = samples.unsqueeze(1) # [B, 1, N, D]
    dist_between_samples = torch.norm(s1 - s2, p=2, dim=-1) # [B, N, N]
    
    # Среднее по всем парам (исключая диагональ для точности)
    num_samples = samples.size(1)
    diversity_term = dist_between_samples.sum(dim=(1, 2)) / (num_samples * (num_samples - 1))

    return dist_to_target - 0.5 * diversity_term

# Пример использования для матчинга:
# Чем меньше score, тем выше соответствие (это "энергия")
vacancy_vector = torch.randn(1, 128)
resume_samples = torch.randn(1, 10, 128) # 10 вариаций вектора навыка
score = energy_score(vacancy_vector, resume_samples)
print(f"Energy Match Score: {score.item()}")

```

### 2. Обучение модели (Energy Loss)

В CALM модель обучается предсказывать такие векторы, которые минимизируют эту энергию. Если вы строите модель матчинга, ваша задача — обучить сеть выдавать векторы резюме, максимально близкие к вакансии в «энергетическом» смысле.

```python
class SkillEnergyModel(nn.Module):
    def __init__(self, dim=128):
        super().__init__()
        self.encoder = nn.Sequential(
            nn.Linear(dim, 256),
            nn.ReLU(),
            nn.Linear(256, dim)
        )

    def forward(self, x):
        return self.encoder(x)

def train_step(model, resume_skills, vacancy_skills, optimizer):
    model.train()
    optimizer.zero_grad()
    
    # Предсказываем латентные векторы (с добавлением шума для вариативности)
    mu = model(resume_skills) 
    # Генерируем несколько сэмплов вокруг предсказания (как в VAE или CALM)
    noise = torch.randn(mu.size(0), 10, mu.size(1)).to(mu.device)
    samples = mu.unsqueeze(1) + 0.1 * noise 
    
    # Считаем Energy Loss
    loss = energy_score(vacancy_skills, samples).mean()
    
    loss.backward()
    optimizer.step()
    return loss.item()

```

### 3. Генерация и Сэмплирование (Langevin Dynamics)

Чтобы «подтянуть» вектор резюме к вакансии или найти идеальный вектор соответствия, в Energy-Based Models (EBM) часто используют градиентный спуск по функции энергии (Langevin Dynamics).

```python
def sample_with_langevin(model, initial_vector, target_vacancy, steps=10, lr=0.1):
    """
    Уточнение вектора навыка для достижения минимальной энергии относительно вакансии
    """
    x = initial_vector.clone().detach().requires_grad_(True)
    
    for _ in range(steps):
        # В данном контексте "модель" может выступать как дискриминатор энергии
        # Здесь мы просто минимизируем дистанцию в энергетическом поле
        samples = x.unsqueeze(1) + 0.01 * torch.randn(x.size(0), 5, x.size(1))
        energy = energy_score(target_vacancy, samples).sum()
        
        energy.backward()
        with torch.no_grad():
            # Двигаемся против градиента энергии + добавляем небольшой шум
            x -= lr * x.grad + 0.005 * torch.randn_like(x)
            x.grad.zero_()
            
    return x

```

### Основные отличия от косинусного сходства:

1. **Устойчивость к выбросам:** Energy Score менее чувствителен к единичным несовпадениям, так как оценивает распределение целиком.
2. **Diversity (Разнообразие):** Вторая часть формулы (`diversity_term`) заставляет систему учитывать разные аспекты навыка, а не просто фокусироваться на самом частотном значении.
3. **Непрерывность:** Вы можете обучать модель в непрерывном пространстве (как в статье CALM), что позволяет делать «мягкий» матчинг даже там, где нет точного совпадения названий навыков.

Этот код можно интегрировать в текущий пайплайн, заменив финальный слой вычисления `cosine_similarity` на функцию `energy_score`. Вам потребуется генерировать несколько эмбеддингов для каждого резюме (например, через Dropout или небольшую вариацию параметров энкодера), чтобы получить `samples`.

Примеры кода можно посмотреть в https://github.com/shaochenze/calm
